x/stakeibc/keeper/msg_server_instant_redeem_stake.go

```diff
@@ -0,0 +1,112 @@
+	sender, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "creator address is invalid: %s. err: %s", msg.Creator, err.Error())
+	}
+
+	// then make sure host zone is valid
+	hostZone, found := k.GetHostZone(ctx, msg.HostZone)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "host zone is invalid: %s", msg.HostZone)
+	}
+
+	// safety check: redemption rate must be within safety bounds
+	rateIsSafe, err := k.IsRedemptionRateWithinSafetyBounds(ctx, hostZone)
+	if !rateIsSafe || (err != nil) {
+		errMsg := fmt.Sprintf("IsRedemptionRateWithinSafetyBounds check failed. hostZone: %s, err: %s", hostZone.String(), err.Error())
+		return nil, sdkerrors.Wrapf(types.ErrRedemptionRateOutsideSafetyBounds, errMsg)
+	}
+
+	// construct desired unstaking amount from host zone
+	nativeAmount := sdk.NewDecFromInt(msg.Amount).Mul(hostZone.RedemptionRate).RoundInt()
+	// Redemption amount must not be greater than staked balance in that zone.
+	if nativeAmount.GT(hostZone.StakedBal) {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, "cannot unstake an amount g.t. staked balance on host zone: %v", msg.Amount)
+	}
+	// safety checks on the coin
+	// 	- Redemption amount must be positive
+	if !nativeAmount.IsPositive() {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "amount must be greater than 0. found: %v", msg.Amount)
+	}
+
+	// get the native Ibc coin to return to user
+	ibcDenom := hostZone.GetIbcDenom()
+	nativeCoin := sdk.NewCoin(ibcDenom, nativeAmount)
+
+	stDenom := types.StAssetDenomFromHostZoneDenom(hostZone.HostDenom)
+	// 	- Creator owns at least "amount" stAssets
+	balance := k.bankKeeper.GetBalance(ctx, sender, stDenom)
+	k.Logger(ctx).Info(fmt.Sprintf("Redemption issuer IBCDenom balance: %v%s", balance.Amount, balance.Denom))
+	if balance.Amount.LT(msg.Amount) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "balance is lower than redemption amount. redemption amount: %v, balance %v: ", msg.Amount, balance.Amount)
+	}
+
+	// Find and subtract this amount from a deposit record if it is big enough
+	depositRecords := k.RecordsKeeper.GetAllDepositRecord(ctx)
+	pendingDepositRecords := utils.FilterDepositRecords(depositRecords, func(record recordstypes.DepositRecord) (condition bool) {
+		return record.Status == recordstypes.DepositRecord_TRANSFER_QUEUE && record.HostZoneId == hostZone.ChainId
+	})
+	totalPendingDeposits := utils.SumDepositRecords(pendingDepositRecords)
+	fmt.Printf("totalPendingDeposits = %v\n", totalPendingDeposits)
+	if nativeAmount.GT(totalPendingDeposits) {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, "cannot fast unbond an amount %v g.t. pending deposit balance on host zone: %v", nativeAmount, msg.Amount)
+	}
+	// Subtract all of nativeAmount from one or more pending deposit records
+	x := nativeAmount
+	for _, depositRecord := range pendingDepositRecords {
```

Comment:
Might not need this loop if we only have one depositRecord matching the conditions, per comment on line 68

Reply:
See comment above.